designs/2024-deprecated-rule-metadata/README.md

```diff
@@ -0,0 +1,250 @@
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think
+    about any opposing viewpoints that reviewers might bring up.
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+There are some limited [backwards compatibility](#backwards-compatibility-analysis) concerns for third-party tooling.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+Existing rules will continue to be backwards-compatible with the new format.
+
+Changing the format of these properties mainly affects third-party documentation tooling and websites that use this information, and not ESLint users nor ESLint plugins directly.
+
+For the most part, the new `meta.deprecated` format should be backwards-compatible, as code is often written to check simply for a truthy value inside of `meta.deprecated`, e.g. `if (rule.meta.deprecated) { /* ... */ }`, which will continue to work as expected. If code checks specifically for the boolean `true` value in `meta.deprecated`, or retrieves rule names from `meta.replacedBy`, then it would need to be updated.
+
+Overall, a limited number of third-party tools that might be affected, and these should be trivial to fix when impacts are discovered.
+
+We do not need to consider this to be a breaking change in terms of ESLint's [semantic versioning policy](https://github.com/eslint/eslint#semantic-versioning-policy).
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+
+1. Do nothing. This would leave the current `meta.deprecated` and `meta.replacedBy` properties as they are, which would continue to be ambiguous and limited in the information they can provide.
+2. Consolidate `meta.replacedBy` into `meta.deprecated.replacedBy`. This alternative was strongly considered. If we were to design things from scratch, we might choose this approach. However, there's not much tangible benefit to it besides organizational tidiness, and the downsides include:
+   - It creates yet another migration burden for the plugin ecosystem (potentially hundreds of plugins) to migrate from the old property to the new property
+   - There's overhead for us to go through the process of deprecating the old property and encouraging or helping plugins to move to the new one
+   - Tooling would need to support both properties, perhaps indefinitely due to the long tail of rules that may never be updated
```

Comment:
[Question] Are there existing examples in the wild of rules that have stayed deprecated for a long time? I'm wondering if there really would be much pain from, say:

1. Adding `meta.deprecated.replacedBy` and marking `meta.replacedBy` as a `@deprecated` alias in ESLint v9
2. Removing `meta.replacedBy` in ESLint v10

The rest of the rule metadata I'd be much more hesitant to change. But [this Sourcegraph search for replacedBy in lint/rule files](https://sourcegraph.com/search?q=context:global+lang:JavaScript+`replacedBy`+file:.*%28lint%7Crule%29.*+-file:.*%28jspm_%7Cnode_modules%29.*+-repo:eslint/eslint+-repo:npm/.*&patternType=keyword&sm=0) shows only a little under 300 code results, with the occasional script.

Reply:
Given that i almost never do major bumps, all of the import, react, and jsx-a11y plugins have examples.

Reply:
It sounds like @JoshuaKGoldberg has a preference for consolidating the properties, whereas @ljharb is neutral or slightly opposed? I'm open to hearing more opinions and still open to changing the proposal on this based on reviewer feedback.

It is true that a migration from `meta.replacedBy` to `meta.deprecated.replacedBy` would be a relatively lightweight and inconsequential migration, compared to more painful ESLint migrations or breaking changes we have performed in the past.

I fleshed out this section with some more pros and cons.

Interested to hear more feedback on this decision.

Reply:
I mean in general I prefer to avoid breaking changes - in particular because then, how will an eslint plugin support both eslint 9 and 10, given that RuleTester fails with unknown properties?

Reply:
@nzakas @mdjermanovic I'm interested to hear if either of you have a preference between keeping the two existing properties or consolidating `meta.replacedBy` into `meta.deprecated.replacedBy`?

Reply:
I'd prefer consolidating `meta.replacedBy` into `meta.deprecated.replacedBy`.

Reply:
:+1: consolidating.